### PR TITLE
Create launcher for multi-target updates

### PR DESCRIPTION
**Pull request by taheris**
Part of [PRO-2972](https://advancedtelematic.atlassian.net/browse/PRO-2972).

### CODE REVIEW

**Comment by heartsucker** [comment]
> +1,882 −281

Always with the huge PRs... 😢

File: sota-launcher/examples/config.toml [hunk 1]
```diff
@@ -0,0 +1,2 @@
+environment = "ci"
+play_session = "PLAY_SESSION=1234567890abcdef1234567890abcdef12345678-id_token=1234567890abcdefghijklmnopqrstuvwxyz.1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABC.1234567890abcdefghijklmnopqrstuvwxyzABCDEFG&access_token=abcdefghijklmnop&auth_plus_access_token=1234567890abcdefghijklmnopqrstuvwxy.1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ1234567890abcdefghijklmn.1234567890abcdefghijklmnopqrstuvwxyzABCDEFG&csrfToken=1234567890abcdef1234567890abcdef12345678-1234567890123-1234567890abcdef12345678&namespace=auth0%7C1234567890abcdefghijklmn"
```
Comment (heartsucker) [inline sota-launcher/examples/config.toml]: I feel like we shouldn't have full session cookie here. Why was this added?

File: sota-launcher/src/datatypes.rs [hunk 2]
```diff
@@ -0,0 +1,234 @@
...
+pub struct Update {
+    pub from: Option<UpdateTarget>,
+    pub to: UpdateTarget
+}
+
+#[allow(non_snake_case)]
+#[derive(Serialize, Deserialize, Debug)]
+pub struct UpdateTarget {
+    pub target: String,
+    pub targetLength: u64,
```
Comment (heartsucker) [inline sota-launcher/src/datatypes.rs]: `#[serde(rename = "targetLength")]` and then you can ditch `#[allow(non_snake_case)]`

File: sota-launcher/src/datatypes.rs [hunk 3]
```diff
@@ -0,0 +1,234 @@
...
+impl Serialize for ChecksumMethod {
+    fn serialize<S: Serializer>(&self, ser: S) -> result::Result<S::Ok, S::Error> {
+        ser.serialize_str(match *self {
+            ChecksumMethod::Sha256 => "sha256",
+            ChecksumMethod::Sha512 => "sha512",
+        })
+    }
+}
+
+impl<'de> Deserialize<'de> for ChecksumMethod {
```
Comment (heartsucker) [inline sota-launcher/src/datatypes.rs]: This can be rewritten

```rust
fn deserialize<D: Deserializer<'de>>(de: D) -> result::Result<Self, D::Error> {
  let s: String = Deserialize::deserialize(de)?;
  s.parse().map_err(|err| serde::de::Error::custom(format!("{}", err)))
}
```

To get the same effect, and then you're not tying yourself to `serde_json`.

But also, you could use `#[serde(rename = "XXX")]` on the variants and not have to have the manual `Serialize`, `Deserialize`, `ToString`, and `FromString` implementations.

File: sota-launcher/src/datatypes.rs [hunk 4]
```diff
@@ -0,0 +1,234 @@
...
+impl Display for PlaySession {
+    fn fmt(&self, f: &mut Formatter) -> fmt::Result {
+        write!(f, "PLAY_SESSION={}&access_token={}&auth_plus_access_token={}&csrfToken={}&namespace={}",
+               self.session, self.access_token, self.auth_plus_token, self.csrf_token, self.namespace)
+    }
+}
+
+impl<'de> Deserialize<'de> for PlaySession {
+    fn deserialize<D: Deserializer<'de>>(de: D) -> result::Result<Self, D::Error> {
+        if let json::Value::String(ref s) = Deserialize::deserialize(de)? {
```
Comment (heartsucker) [inline sota-launcher/src/datatypes.rs]: Similar note about the `serde` impls as above.

File: sota-launcher/src/main.rs [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+    let mtu = MultiTargetUpdate::new(&config)?;
+    let update_id = mtu.create(&updates)?;
+    info!("update_id: {}", update_id);
+    mtu.launch(targets.device.device_id, update_id)?;
+
+    Ok(())
+}
+
+fn parse_args() -> Result<(Config, Targets)> {
+    let matches = clap_app!(
```
Comment (heartsucker) [inline sota-launcher/src/main.rs]: God, rust macros are amazing.

File: sota-launcher/examples/config.toml (hunk 1 above)
Comment (heartsucker) [inline sota-launcher/examples/config.toml]: Also, the fact that our session cookies aren't encrypted: so sad.

Comment (taheris) [inline sota-launcher/examples/config.toml]: Because our web API expects it. Extracting it every hour or so really sucks and I'd be happy to get rid of it entirely if there's another way.

File: sota-launcher/src/datatypes.rs [hunk 6]
```diff
@@ -60,11 +59,8 @@ impl Display for Environment {
 
 impl<'de> Deserialize<'de> for Environment {
     fn deserialize<D: Deserializer<'de>>(de: D) -> result::Result<Self, D::Error> {
-        if let json::Value::String(ref s) = Deserialize::deserialize(de)? {
-            s.parse().map_err(|err| serde::de::Error::custom(format!("{}", err)))
-        } else {
-            Err(serde::de::Error::custom("expected string type for environment"))
-        }
+        let s: String = Deserialize::deserialize(de)?;
```
Comment (heartsucker) [inline sota-launcher/src/datatypes.rs]: :D

File: sota-launcher/examples/config.toml (hunk 1 above)
Comment (heartsucker) [inline sota-launcher/examples/config.toml]: Fair enough. I feel like the CSRF token could expire, so this might have to be automated eventually, but that can be dealt with eventually.

File: sota-launcher/src/datatypes.rs (hunk 3 above)
Comment (taheris) [inline sota-launcher/src/datatypes.rs]: Thanks for the suggestions. I've updated it to the first way. I tried the rename method as well but it seems the rename field attribute isn't supported on enums.

Comment (heartsucker) [inline sota-launcher/src/datatypes.rs]: I was lied to! https://serde.rs/variant-attrs.html

Comment (heartsucker) [inline sota-launcher/src/datatypes.rs]: But to be fair, I've never used this feature, so idk how it actually works.
